experimental/displaynames/src/displaynames.rs

```diff
@@ -399,90 +453,184 @@ impl LocaleDisplayNamesFormatter {
     );
 
     /// Returns the display name of a locale.
+    /// This implementation is based on the algorithm described in
+    /// https://www.unicode.org/reports/tr35/tr35-general.html#locale_display_name_algorithm
+    ///
     // TODO: Make this return a writeable instead of using alloc
     pub fn of<'a, 'b: 'a, 'c: 'a>(&'b self, locale: &'c Locale) -> Cow<'a, str> {
-        // https://www.unicode.org/reports/tr35/tr35-general.html#Display_Name_Elements
-
-        // TODO: This binary search needs to return the longest matching found prefix
-        // instead of just perfect matches
-        if let Some(displayname) = match self.options.style {
-            Some(Style::Short) => self
-                .locale_data
-                .get()
-                .short_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            Some(Style::Long) => self
-                .locale_data
-                .get()
-                .long_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            Some(Style::Menu) => self
-                .locale_data
-                .get()
-                .menu_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            _ => None,
-        }
-        .or_else(|| {
-            self.locale_data
-                .get()
-                .names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse())
-        }) {
-            return Cow::Borrowed(displayname);
+        let longest_matching_subtag =
+            LongestMatchingSubtag::find_longest_matching_subtag(&locale, &self);
+
+        // Step - 1: Construct a locale display name string (LDN).
+        // Find the displayname for the longest_matching_subtag which was derived above.
+        let ldn = get_locale_display_name(&locale, longest_matching_subtag, &self);
+
+        // Step - 2: Construct a vector of longest qualifying substrings (LQS).
+        // Find the displayname for the remaining locale if exists.
+        let lqs = get_longest_qualifying_substrings(&locale, longest_matching_subtag, &self);
+
+        // Step - 3: Return the displayname based on the size of LQS.
+        let mut result = Cow::Borrowed(ldn);
+        if lqs.len() > 0 {
+            let mut output = String::with_capacity(
+                result.len() + " (".len() + lqs.iter().map(|s| ", ".len() + s.len()).sum::<usize>()
+                    - ", ".len()
+                    + ")".len(),
```

Comment:
These strings should come from locale data. Could be a follow-up but we definitely need this before this goes stable.

https://github.com/unicode-org/cldr-json/blob/80a94b0f6c3a34d6e2dc0dca8639a54babc87f94/cldr-json/cldr-localenames-full/main/zh/localeDisplayNames.json#L12C1-L14C44

Reply:
Acknolwedged

Reply:
#3913